src/NHibernate/Hql/Ast/ANTLR/HqlSqlWalker.cs

```diff
@@ -98,6 +99,21 @@ public override void ReportError(RecognitionException e)
 			_parseErrorHandler.ReportError(e);
 		}
 
+		internal IStatement Transform()
+		{
+			var tree = (IStatement) statement().Tree;
+			// Use the guessed type in case we weren't been able to detect the type
+			foreach (var parameter in _parameters)
+			{
+				if (parameter.ExpectedType == null && _guessedParameterTypes.TryGetValue(parameter, out var guessedType))
+				{
+					parameter.ExpectedType = guessedType;
```

Comment:
Why do we need `_guessedParameterTypes` dictionary? Why we cannot just do this: 

```csharp
if (parameter.ExpectedType == null && parameter.GuessedType != null)
{
    parameter.ExpectedType = parameter.GuessedType;
}
```

Reply:
Answering my own question: because `parameter` here does not have `GuessedType` property.